### PR TITLE
Throw on invalid env key

### LINKED ISSUE #69: Invalid ENV declaration should return an error
If I do the following in my `build.earth` file:

```
ENV ${CODE_SRC}=/go/src/<project>

protobuf:
    COPY --dir api ${CODE_SRC}/backend
```

Then running `earth +protobuf` results in the files being copied into `/backend`:

`+protobuf | --> COPY [api] /backend`

I have worked around this by specifying the full path in the copy command:

```
ENV ${CODE_SRC}=/go/src/<project>

protobuf:
    COPY --dir api /go/src/<project>/backend
```

### PR DESCRIPTION
**Pull request by drews256**
Cool project. I was looking for something like this recently, wish I had seen it before I found a different solution.

Here is an attempt to fix #69 

I used regex to match invalid characters. It's specific to the error seen in #69. It matches against `[]{}$` characters. Though the more I think about it maybe it should be the other way around and ensure a match on uppercase A-Z. 

I figured I'd open this guy up first.

I have a log statement in there that should be taken out if this is the path you wanna go down.

### CODE REVIEW

File: earthfile2llb/converter.go [hunk 1]
```diff
@@ -477,14 +478,21 @@ func (c *Converter) Volume(ctx context.Context, volumes []string) {
 }
 
 // Env applies the ENV command.
-func (c *Converter) Env(ctx context.Context, envKey string, envValue string) {
+func (c *Converter) Env(ctx context.Context, envKey string, envValue string) error {
 	envValue = c.expandArgs(envValue)
+	invalidEnvKey, err := regexp.MatchString(`[${}()]`, envKey)
```
Comment (vladaionescu) [inline earthfile2llb/converter.go:483]: In would go for `[a-zA-Z_]+[a-zA-Z0-9_]*` as you suggested in the PR description. [Source](https://stackoverflow.com/a/2821201).

File: earthfile2llb/converter.go [hunk 2]
```diff
@@ -477,14 +478,21 @@ func (c *Converter) Volume(ctx context.Context, volumes []string) {
...
+	logging.GetLogger(ctx).With("env-key", envKey).With("env-key-value", invalidEnvKey).Info("ENV Key Value")
+	if invalidEnvKey || err != nil {
+		logging.GetLogger(ctx).Info("Failed to apply ENV")
+		return fmt.Errorf("Invalid env key definition %s", envKey)
+	}
```
Comment (vladaionescu) [inline earthfile2llb/converter.go:488]: Can you move this logic into `listener.go`? (ie the caller of this)

The overall idea is to put any syntax related checks before it gets to the converter.

Comment (vladaionescu) [inline earthfile2llb/converter.go:488]: Side note: No need to do logging - just make sure `l.err` is set after you move to listener.

File: go.mod [hunk 3]
```diff
@@ -23,8 +23,12 @@ require (
...
 	golang.org/x/crypto v0.0.0-20200302210943-78000ba7a073
 	golang.org/x/lint v0.0.0-20190930215403-16217165b5de // indirect
-	golang.org/x/sync v0.0.0-20190911185100-cd5d95a43a6e
-	golang.org/x/xerrors v0.0.0-20191011141410-1b5146add898 // indirect
+	golang.org/x/mod v0.3.0 // indirect
+	golang.org/x/sync v0.0.0-20200317015054-43a5402ce75a
+	golang.org/x/tools v0.0.0-20200515220128-d3bf790afa53 // indirect
+	golang.org/x/tools/gopls v0.4.1 // indirect
+	honnef.co/go/tools v0.0.1-2020.1.4 // indirect
+	mvdan.cc/xurls/v2 v2.2.0 // indirect
```
Comment (vladaionescu) [inline go.mod:31]: These dependency changes seem unrelated? Mind reverting these? (Unless I'm missing something)

**Comment by drews256** [comment]
@vladaionescu here's another option addressing the comments you made and with the regex you shared.

File: earthfile2llb/listener.go [hunk 4]
```diff
@@ -595,7 +595,12 @@ func (l *listener) EnterEnvArgKey(c *parser.EnvArgKeyContext) {
 	if l.shouldSkip() {
 		return
 	}
-	l.envArgKey = c.GetText()
+	var envKey = c.GetText()
+	var matchedEnvKey = checkShellName(envKey)
+	if matchedEnvKey != envKey {
+		l.err = fmt.Errorf("Invalid env key definition %s", envKey)
```
Comment (vladaionescu) [inline earthfile2llb/listener.go:601]: Best to add a return so that nothing else is applied.

```suggestion
		l.err = fmt.Errorf("Invalid env key definition %s", envKey)
		return
```

File: earthfile2llb/listener.go [hunk 5]
```diff
@@ -662,6 +667,12 @@ func (ssf *StringSliceFlag) Set(arg string) error {
 	return nil
 }
 
+var shellNameStringRegexp = regexp.MustCompile("^[a-zA-Z_]+[a-zA-Z0-9_]*")
+
+func checkShellName(str string) string {
```
Comment (vladaionescu) [inline earthfile2llb/listener.go:672]: Mind renaming this to `checkEnvVarName`? "Shell name" could be misunderstood as the shell itself.

File: earthfile2llb/listener.go [hunk 6]
```diff
@@ -662,6 +667,12 @@ func (ssf *StringSliceFlag) Set(arg string) error {
...
+	return shellNameStringRegexp.FindString(str)
```
Comment (vladaionescu) [inline earthfile2llb/listener.go:673]: Can you make this func return a `bool`? Or even better - an `error`? You'll probably need to use `regexp.MatchString`.

File: earthfile2llb/listener.go [hunk 7]
```diff
@@ -662,6 +667,12 @@ func (ssf *StringSliceFlag) Set(arg string) error {
 	return nil
 }
 
+var shellNameStringRegexp = regexp.MustCompile("^[a-zA-Z_]+[a-zA-Z0-9_]*")
```
Comment (vladaionescu) [inline earthfile2llb/listener.go:670]: Best to also end with a `$` to ensure that the entire strings matches.

**Comment by alexcb** [comment]
We are following github's push to rename `master` to `main`, see https://github.com/github/renaming/ for details.

If this PR is ready to be merged, please open a new PR against our `main` branch, as we no longer have a `master` branch. Thanks.
